### PR TITLE
Rename "bucket" to "gcs_bucket" in GCSToS3Operator

### PR DESCRIPTION
**Pull request by hankehly**
Related: #32760

This pull request renames `bucket` to `gcs_bucket` on the `GCSToS3Operator` class to distinguish between AWS and GCP buckets. The change supports both parameter names but issues a deprecation warning for `bucket`.

Here are the corresponding unit/system tests:
```shell
# Unit tests
pytest tests/providers/amazon/aws/transfers/test_gcs_to_s3.py

# System tests (fill out the below env vars beforehand)
export AWS_ACCESS_KEY_ID=***
export AWS_SECRET_ACCESS_KEY=***
export AWS_DEFAULT_REGION=***
export GOOGLE_APPLICATION_CREDENTIALS=/path/to/credentials.json
export GCP_PROJECT_ID=***
pytest --system amazon tests/system/providers/amazon/aws/example_gcs_to_s3.py
```

Your review is appreciated.

### CODE REVIEW

File: airflow/providers/amazon/aws/transfers/gcs_to_s3.py [hunk 1]
```diff
@@ -117,7 +120,21 @@ def __init__(
     ) -> None:
         super().__init__(**kwargs)
 
-        self.bucket = bucket
+        if bucket and gcs_bucket:
```
Comment (hankehly) [inline airflow/providers/amazon/aws/transfers/gcs_to_s3.py]: `bucket` is still accepted, but outputs a deprecation warning.

File: tests/providers/amazon/aws/transfers/test_gcs_to_s3.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ def test_execute__match_glob(self, mock_hook):
 
             operator = GCSToS3Operator(
                 task_id=TASK_ID,
-                bucket=GCS_BUCKET,
+                gcs_bucket=GCS_BUCKET,
```
Comment (hankehly) [inline tests/providers/amazon/aws/transfers/test_gcs_to_s3.py:60]: All unit tests (except for the one that checks for backwards compatibility) use the updated naming.

**Comment by hankehly** [comment]
@ferruzzi This is the PR we briefly talked about in [this other issue](https://github.com/apache/airflow/pull/32760#discussion_r1276725424). I would appreciate your review.

File: airflow/providers/amazon/aws/transfers/gcs_to_s3.py [hunk 3]
```diff
@@ -88,6 +89,7 @@ class GCSToS3Operator(BaseOperator):
 
     template_fields: Sequence[str] = (
         "bucket",
+        "gcs_bucket",
```
Comment (eladkal) [inline airflow/providers/amazon/aws/transfers/gcs_to_s3.py:91]: now that we have `gcs_bucket` we can safely remove `bucket` from the list dont we?

Comment (hankehly) [reply airflow/providers/amazon/aws/transfers/gcs_to_s3.py:91]: Yes. I see now the "rendering" of templated fields occurs on `execute()` so no need to include the deprecated param here. Removed in 03c4bc8. Thanks.

File: airflow/providers/amazon/aws/transfers/gcs_to_s3.py [hunk 4]
```diff
@@ -41,7 +41,8 @@ class GCSToS3Operator(BaseOperator):
         For more information on how to use this operator, take a look at the guide:
         :ref:`howto/operator:GCSToS3Operator`
 
-    :param bucket: The Google Cloud Storage bucket to find the objects. (templated)
+    :param gcs_bucket: The Google Cloud Storage bucket to find the objects. (templated)
+    :param bucket: (Deprecated) Use gcs_bucket instead.
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/transfers/gcs_to_s3.py]:
```suggestion
    :param bucket: (Deprecated) Use ``gcs_bucket`` instead.
```

File: airflow/providers/amazon/aws/transfers/gcs_to_s3.py [hunk 5]
```diff
@@ -117,7 +119,23 @@ def __init__(
     ) -> None:
         super().__init__(**kwargs)
 
-        self.bucket = bucket
+        if bucket and gcs_bucket:
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/transfers/gcs_to_s3.py]: Do we need this check? If both are provided, we should use `gcs_bucket` IMO. To me we should do (pseudo code):

```python
if bucket:
  warnings.warn()
  self.gcs_bucket = bucket

if gcs_bucket:
  self.gcs_bucket = gcs_bucket
```

As such, if both are provided, `gcs_bucket` is used

File: airflow/providers/amazon/aws/transfers/gcs_to_s3.py [hunk 6]
```diff
@@ -117,7 +119,23 @@ def __init__(
...
+                stacklevel=2,
+            )
+            self.gcs_bucket = bucket
+        elif gcs_bucket:
+            self.gcs_bucket = gcs_bucket
+        else:
+            raise ValueError("You must pass either bucket or gcs_bucket.")
+        # The bucket attribute is unused, but must be set on the operator instance in
+        # order to include it in template_fields.
+        self.bucket = None
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/transfers/gcs_to_s3.py]: No longer needed?

File: tests/providers/amazon/aws/transfers/test_gcs_to_s3.py (hunk 2 above)
Comment (vincbeck) [reply tests/providers/amazon/aws/transfers/test_gcs_to_s3.py:60]: Nice :)

File: tests/providers/amazon/aws/transfers/test_gcs_to_s3.py [hunk 7]
```diff
@@ -196,6 +196,50 @@ def test_execute(self, mock_hook):
             assert sorted(MOCK_FILES) == sorted(uploaded_files)
             assert sorted(MOCK_FILES) == sorted(hook.list_keys("bucket", delimiter="/"))
 
+    @mock.patch("airflow.providers.amazon.aws.transfers.gcs_to_s3.GCSHook")
+    def test_execute_gcs_bucket_rename_compatibility(self, mock_hook):
```
Comment (vincbeck) [inline tests/providers/amazon/aws/transfers/test_gcs_to_s3.py:200]: Love that

File: airflow/providers/amazon/aws/transfers/gcs_to_s3.py [hunk 8]
```diff
@@ -117,7 +119,18 @@ def __init__(
...
+                "The 'bucket' parameter is deprecated and will be removed in a future version. "
+                "Please use 'gcs_bucket' instead.",
+                AirflowProviderDeprecationWarning,
+                stacklevel=2,
+            )
+            self.gcs_bucket = bucket
+        elif gcs_bucket:
+            self.gcs_bucket = gcs_bucket
+        else:
+            raise ValueError("You must pass either bucket or gcs_bucket.")
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/transfers/gcs_to_s3.py]:
```suggestion
            raise ValueError("You must pass either ``bucket`` or ``gcs_bucket``.")
```

File: airflow/providers/amazon/aws/transfers/gcs_to_s3.py [hunk 9]
```diff
@@ -117,7 +119,18 @@ def __init__(
     ) -> None:
         super().__init__(**kwargs)
 
-        self.bucket = bucket
+        if bucket:
+            warnings.warn(
+                "The 'bucket' parameter is deprecated and will be removed in a future version. "
+                "Please use 'gcs_bucket' instead.",
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/transfers/gcs_to_s3.py]:
```suggestion
                "The ``bucket`` parameter is deprecated and will be removed in a future version. "
                "Please use ``gcs_bucket`` instead.",
```

File: airflow/providers/amazon/aws/transfers/gcs_to_s3.py [hunk 10]
```diff
@@ -117,7 +119,18 @@ def __init__(
...
+                AirflowProviderDeprecationWarning,
+                stacklevel=2,
+            )
+            self.gcs_bucket = bucket
+        elif gcs_bucket:
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/transfers/gcs_to_s3.py]: By just doing a if, if both are provided, we use `gcs_bucket`, which is what we want

```suggestion
        if gcs_bucket:
```

Comment (hankehly) [reply airflow/providers/amazon/aws/transfers/gcs_to_s3.py]: Sorry, I missed that in your previous comment.

**Review by vincbeck — APPROVED** [review]

**Pull request merged**